File: Rosu/programmazione/lezioni/java/eratosthenes/src/ro/razvan/uniud/eratosthenes/java/Eratosthenes.java
package ro.razvan.uniud.eratosthenes.java;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

final class Eratosthenes {

    /**
     * Calcola tutti i numeri primo fino a n
     * Pre: n > 1
     *
     * @param n intero maggiore di 1
     * @return lista di interi primi minori o uguali a n
     */
    @NotNull
    static List<Integer> primesUpTo(final int n) {

        final var primes = new ArrayList<Integer>();

        final Boolean[] sieve = IntStream.range(0, n + 1)
                .mapToObj((element) -> element >= 2)
                .toArray(Boolean[]::new);

        for (int p = 2; p < sieve.length; p++) {

            if (sieve[p]) {

                primes.add(p);

                for (int i = p * 2; i < sieve.length; i += p) {
                    sieve[i] = false;
                }

            }

        }

        return primes;
    }

    public static void main(@NotNull final String[] args) {
        println(primesUpTo(Integer.parseInt(args[0])));
    }

    static void println(@Nullable final Object any) {
        System.out.println(any);
    }

}

File: Rosu/programmazione/lezioni/java/eratosthenes/src/ro/razvan/uniud/eratosthenes/kotlin/Eratosthenes.kt
package ro.razvan.uniud.eratosthenes.kotlin


object Eratosthenes {

    /**
     * Calcola tutti i numeri primo fino a [n]
     * Pre: [n] > 1
     *
     * @return tutti i numeri primi minori o uguali a [n]
     */
    fun primesUpTo(n: Int): List<Int> {

        val sieve = BooleanArray(n + 1) { i ->
            i >= 2
        }
        val primes = ArrayList<Int>()
        var p = 2

        /* Term: n + 1 - p */
        while (p <= n) {

            if (sieve[p]) {

                primes.add(p)
                var i = 2 * p

                while (i <= n) {
                    sieve[i] = false
                    i += p
                }

            }

            p++
        }

        return primes // Post: primes = { primi <= n }
    }

    @JvmStatic
    fun main(args: Array<String>) {
        println(primesUpTo(args[0].toInt()))
    }

}